ui/src/router.tsx

```diff
@@ -0,0 +1,94 @@
+import Button from "@mui/material/Button";
+import Edit from "edit";
+import { PathError } from "hooks";
+import Overview from "overview";
+import { ErrorBoundary } from "react-error-boundary";
+import {
+  Route,
+  Switch,
+  useHistory,
+  useParams,
+  useRouteMatch,
+} from "react-router-dom";
+import { UnderlaySelect } from "underlaySelect";
+import { Datasets } from "./datasets";
+
+const pages = {
+  "/": UnderlaySelect,
+  "/:underlayName": Datasets,
+  "/:underlayName/cohorts/:cohortId": Overview,
+  "/:underlayName/cohorts/:cohortId/edit/:groupId/:criteriaId": Edit,
+};
+
+export function AppRouter() {
+  return (
+    <ErrorBoundary FallbackComponent={NotFound}>
+      <Switch>
+        {Object.entries(pages).map(([page, component]) => (
+          <Route exact key={page} path={page} component={component} />
+        ))}
+        <Route path="/">
+          <NotFound error={new PathError("Invalid URL.")} />
+        </Route>
+      </Switch>
+    </ErrorBoundary>
+  );
+}
+
+// useBackUrl returns the second longest matching page, assuming that the
+// longest is the current page since longer ones wouldn't match.
+export function useBackUrl(): string | undefined {
+  const page = Object.keys(pages)
+    .filter((page) => useRouteMatch({ path: page }))
```

Comment:
This breaks the first rule of hooks: https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

Reply:
This usage is okay because it's iterating over a constant array. The order and number of hooks never changes.

Reply:
I would say it still violates the rule, even if it doesn't break the underlying premise. In terms of readability, I think this also requires too much derivation to conclude that it's safe.

If there's no other way of implementing this without hooks, then I think at least a comment here is warranted.

Reply:
Done. I found a way to implement it without without using a hook in the loop.